statie/source/_posts/2019/2019-06-02-logic-belongs-to-entities.md

```diff
@@ -0,0 +1,258 @@
+```
+
+As we can see, the method is logically as same as it was in `CartTotalPriceService`.
+But there are a couple of important differences
+* The logic sits together with data in the Cart entity, this leads to a rich model
+* The Cart doesn't expose data (Items), Items are therefore properly encapsulated
+* This code is testable as we can mock the interface in tests to produce predictable prices
+* By the way, this was our goal - to have the logic in the entity
+
+### Interface Implementation
+
+The interface has to be implemented anyway, but that's not a problem at all
+```php
+class PDOProductPrices implements ProductPrices {
+    /**
+     * @var PDO
+     */
+    private $pdo;
+    
+    // constructor
+    
+    public function refreshPrice(int $productId): Money {
+        $statement = $this->pdo->query("SELECT price FROM product WHERE id = :id");
+        $statement->bindParam(':id', $productId);
+        $result = $statement->fetch(PDO::FETCH_ASSOC);
+        $number = (string) $result['price'];
+        return Money::fromString($number);
+    }
+}
+```
+
+Do you remember `CartTotalPriceService`? The code is the same.
+
+And again we can use our favorite ORM, or database layer to implement the interface.
+We can do even more, the implementation can use different system module or even API to a different system.
+We are not limited by technology, all thanks to the interface.
+
+#### Only One Implementation
+
+We will usually have only one implementation for such an interface, that is strange, isn't it wrong?
+
+We are using interface to separate layers in this case.
+We separate the business logic layer and infrastructure (database) layer.
+And that is a valid interface usage.
+
+## Moving Logic ...for Maniacs
+
+Do you think the logic is moved well?
+I have a proposal for maniacs.
+
+An Item is probably a value object an moving the logic into value objects is even greater than moving the logic to entities.
+That's because value objects are immutable and the logic is, therefore, super easily testable and super separated.
+
+So let's move some logic into the Item
+```php
+class Item {
+    /**
+     * @var int
+     */
+    private $productId;
+
+    /**
+     * @var float
+     */
+    private $amount;
+
+    // constructor
+
+    public function calculatePrice(ProductPrices $productPrices): Money {
+        $price = $productPrices->refreshPrice($this->productId);
+        return $price->multiply($this->amount);
+    }
+}
+
+class Cart {
+    /**
+     * var Item[]
+     */
+    private $items;
+    
+    // constructor
+
+    public function calculateTotalPrice(ProductPrices $productPrices): Money {
+        $prices = [];
+        foreach ($this->items as $item) {
+            $prices[] = $item->calculatePrice($productPrices);
+        }
+        
+        return Money::sum($prices);
+    }
+}
+
+// usage is still the same
+$totalPrice = $cart->calculateTotalPrice($this->productPrices);
```

Comment:
So basically you pass `ProductPrices` service as entity method argument and itterate over to it?

That creates circular dependencies in my thinking, saying that it's ok that "any object can access any service, if passes as argument".

Personally I prefer tree structure - each object has less responsibility the further we go: 

- ↓ contrainer (has access to all services)
- ↓ service (can manipulate any object)
- ↓ entity (can work with inner data it already contains)
- ↓ value object (can return data)
- ↓ scalars - string, int... (contains data)

Nothing to argue about related to entities and logic, because I don't have any ideas about it.
Just sharing my way of thinkink that makes me fast and flexible with strict clear rules.

Reply:
Well I don't consider `ProductPrices` to be a service. It doesn't even fit your service definition "manipulate any object".

`ProductPrices` is a domain element responsible for "refreshing price". I don't care weather it is an object or an interface, it's just a domain element. I also don't care how is the interface implemented - it could be implemented as independent object (your entity or value object) or as object with dependencies (your service?).

Back to domain elements - I consider following elements on the same level
* entity
* value object
* collection (aka repository)
* interface that represent something in the domain

These elements can have temporary dependencies on any other element without any restrictions. The dependency must be of course reasonable. Dependencies are temporary, so there are no circular dependencies (meaning `new A($b), new B($a)`).

This style allows me to capture the whole business use-case in a single layer.

Reply:
I'd now update my answer with:
 
- ↓ service (can change any object/get data from not itself)